File: LeetCode/Python Solutions/Path With Minimum Effort/path.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        # Define movement directions.
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        # Check if we can reach the destination with a given effort.
        def canReachDestination(maxEffort):
            # Initialize visited cells.
            visited = [[False for _ in range(len(heights[0]))] for _ in range(len(heights))]
            
            # Start DFS using a stack from the top-left corner.
            stack = [(0, 0)]
            visited[0][0] = True
            
            # While there are cells to explore:
            while stack:
                i, j = stack.pop()
                
                # If we're at the destination, return True.
                if i == len(heights) - 1 and j == len(heights[0]) - 1:
                    return True

                # Explore all possible directions.
                for x, y in dirs:
                    ni, nj = i + x, j + y
                    
                    # If the new cell is valid, mark it as visited and add to the stack.
                    if 0 <= ni < len(heights) and 0 <= nj < len(heights[0]) and not visited[ni][nj]:
                        if abs(heights[i][j] - heights[ni][nj]) <= maxEffort:
                            visited[ni][nj] = True
                            stack.append((ni, nj))

            # Return False if we cannot reach the destination.
            return False

        # Initialize binary search range.
        left, right = 0, 1000000
        
        # Binary search logic.
        while left < right:
            mid = (left + right) // 2
            if canReachDestination(mid):
                right = mid
            else:
                left = mid + 1

        # Return the minimum effort.
        return left

File: LeetCode/Java Solutions/Path With Minimum Effort/path.java
class Solution {

    // Define possible movement directions - up, down, left, right.
    private int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    public int minimumEffortPath(int[][] heights) {

        // Initialize the search range for our binary search.
        int left = 0, right = 1000000;

        // Binary search: Continue until the range narrows down to a single value.
        while (left < right) {

            // Calculate the middle effort.
            int mid = (left + right) / 2;

            // Check if it's possible to reach the destination with the current effort.
            if (canReachDestination(heights, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        // The minimum effort required.
        return left;
    }

    // Helper function to check if we can reach the destination with the given effort.
    private boolean canReachDestination(int[][] heights, int maxEffort) {

        // Mark cells as visited or not.
        boolean[][] visited = new boolean[heights.length][heights[0].length];

        // Start DFS from the top-left corner.
        return dfs(heights, 0, 0, maxEffort, visited);
    }

    // Depth-First Search to explore the grid.
    private boolean dfs(int[][] heights, int i, int j, int maxEffort, boolean[][] visited) {

        // If we have reached the destination, return true.
        if (i == heights.length - 1 && j == heights[0].length - 1) return true;

        // Mark the current cell as visited.
        visited[i][j] = true;

        // Explore each direction.
        for (int[] dir : directions) {
            int x = i + dir[0], y = j + dir[1];

            // If the new cell is within bounds, not visited, and the absolute difference in 
            // heights is within the allowed effort, explore it.
            if (x >= 0 && x < heights.length && y >= 0 && y < heights[0].length && 
                !visited[x][y] && Math.abs(heights[i][j] - heights[x][y]) <= maxEffort) {
                if (dfs(heights, x, y, maxEffort, visited)) return true;
            }
        }

        // Return false if we cannot reach the destination.
        return false;
    }
}
